### PR TITLE
feat(router): Add ability to return UrlTree with `NavigationBehaviorOptions` from guards

### LINKED ISSUE #17004: Render 404 page with CanActivate guard without aborting navigation
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Background**
When a user navigates to a route that requires authentication and they are not authenticated, there's no way to use a `CanActivate` guard to render a 404 component and still update the url to what the user was trying to access. 

**Current behavior**
Let's say a user navigates to `/protected` and is not authenticated. Currently, in the `CanActivate` guard, in order to render my 404 component, I would use `router.navigate(['404'], { skipLocationChange: true }); return false;`. However, after rendering the 404 component, the url in the address bar is the previous url before the navigation since the navigation was aborted.

**Expected behavior**
In `CanActivate` guard, there should be a way to render the 404 route while still updating the url in the address bar to `/protected`.

**Minimal reproduction of the problem with instructions**
```ts
@Injectable()
export class LoggedInGuard implements CanActivate {
  constructor(private router: Router, private userService: UserService) {}

  canActivate() {
    if (this.userService.isLoggedIn) return true;

    this.router.navigate(['/404'], { skipLocationChange: true });
    return false;
  }
}
```
This code will render a 404 page, but not update the url. I would like it to render 404 and update the url in the address bar to what the user was trying to access.

**What is the motivation / use case for changing the behavior?**
Github does this, for example, if you go to https://github.com/simonxca/webpack-guide, it will render a 404 page even though the repo exists because you're not authenticated. The motivation is to prevent users from being able to distinguish between a protected from a non-existent page for security.

**Please tell us about your environment:**
Mac OSX Sierra, Node 6.10.3

* **Angular version:** 4.1.X

* **Browser:** all
 
* **Language:** all

### LINKED ISSUE #27148: NavigationExtras when returning URLTree from guard
# 🚀 feature request

### Relevant Package
This feature request is for @angular/router

### Description
In #26521 (Angular 7.1.0), a feature was introduced that allows returning a `URLTree` from guards in order to initiate a redirect. However, this doesn't seem to allow using `NavigationExtras` such as `skipLocationChange`.

Since the change in #26521 wasn't just syntax sugar, but is relevant to have predictable behavior when dealing with multiple guards, it should be considered to somehow allow using these options when returning a `URLTree` as well.

### Describe the solution you'd like
I don't have a really good proposal at this moment apart from next to returning `URLTree` also returning some kind of

```
// TODO: Find a better name?
interface NavigationOptions {
  urlTree: URLTree;
  navigationExtras: NavigationExtras;
}
```

such that I could use

```
class AwesomeGuard implements CanActivate {
  constructor(private router: Router) {}

  canActivate(): NavigationOptions | Observable<NavigationOptions> {
    return {
      urlTree: this.router.parseUrl("…"),
      navigationExtras: {
        skipLocationChange: true,
      },
    };
  }
}
```

I guess the exact typing of the guard would have to be

```
// Just a placeholder name
type GuardReturnType = boolean | URLTree | NavigationOptions;

interface CanActivate {
  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): 
    GuardReturnType 
    | Promise<GuardReturnType> 
    | Observable<GuardReturnType>
}
```

Admittedly, the return type of this function is growing increasingly complex with this.

### Describe alternatives you've considered
Possible options that I don't think would make good solutions would be

* Returning an array `[URLTree, NavigationExtras]` (this just feels unclean)
* Extending URLTree with the navigationExtras and extend parseUrl to take them as well (this muddies the entire API for a change in a specific location)

### PR DESCRIPTION
**Pull request by atscott**
Returning `UrlTree` from a guard was a convenient new feature added to
the `Router`. However, it does not have feature-parity with the old
`router.navigate(...); return false;` pattern. The most common use-case
for this feature is to redirect to a new page _without_ updating the URL
from the initially attempted navigation. For example, rendering a 404
page when the user does not have access privelages to a route.

Fixes #27148

### CODE REVIEW

**Comment by atscott** [comment]
Results from running migration the migration internally: [cl/427481595](http://cl/427481595)

**Comment by atscott** [comment]
[presubmit](http://test/OCL:427493712:BASE:427550812:1644440654334:662fc0f2)

File: goldens/public-api/router/router.md [hunk 1]
```diff
@@ -843,6 +846,14 @@ export class UrlTree {
     toString(): string;
 }
 
+// @public (undocumented)
+export interface UrlTreeWithOptions {
```
Comment (atscott) [inline goldens/public-api/router/router.md:850]: Happy to bikeshed the name of this interface at this time. I'm thinking some other good options might be `RedirectingGuardResult` / `RedirectCommand` / `GuardRedirectCommand` / `GuardRedirect`

File: packages/router/src/router.ts [hunk 2]
```diff
@@ -799,13 +749,13 @@ export class Router {
...
-                       if (isUrlTree(t.guardsResult)) {
-                         const error: Error&{url?: UrlTree} = navigationCancelingError(
-                             `Redirecting to "${this.serializeUrl(t.guardsResult)}"`);
-                         error.url = t.guardsResult;
-                         throw error;
+                       if (t.guardsResult && typeof t.guardsResult !== 'boolean') {
+                         const {urlTree, navigationBehaviorOptions} = isUrlTree(t.guardsResult) ?
+                             {urlTree: t.guardsResult, navigationBehaviorOptions: undefined} :
+                             t.guardsResult;
+                         throw redirectingNavigationError(
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:756]: This logic is repeated twice (both times when `redirectingNavigationError` is used). May be pull it into the `redirectingNavigationError` iteself?

File: packages/router/src/apply_redirects.ts [hunk 3]
```diff
@@ -377,13 +378,14 @@ class ApplyRedirects {
...
-              if (!isUrlTree(result)) return;
-
-              const error: Error&{url?: UrlTree} = navigationCancelingError(
-                  `Redirecting to "${this.urlSerializer.serialize(result)}"`);
-              error.url = result;
-              throw error;
+            tap((result: GuardResult) => {
+              if (typeof result === 'boolean') return;
+
+              const {urlTree, navigationBehaviorOptions} = isUrlTree(result) ?
```
Comment (AndrewKushnir) [inline packages/router/src/apply_redirects.ts:384]: Just as an idea: should we add an `ngDevMode`-guarded check and `console.warn` a message if the `result` has an old (previous) format that would be deprecated?

File: packages/router/src/models.ts [hunk 4]
```diff
@@ -12,6 +12,85 @@ import {Observable} from 'rxjs';
 import {ActivatedRouteSnapshot, RouterStateSnapshot} from './router_state';
 import {UrlSegment, UrlSegmentGroup, UrlTree} from './url_tree';
 
+/**
+ * @publicApi
+ */
+export type GuardResult = boolean|UrlTreeWithOptions|DeprecatedGuardResult;
```
Comment (AndrewKushnir) [inline packages/router/src/models.ts:18]: I think we'd need to add docs to the `GuardResult`, `UrlTreeWithOptions` and `MaybeAsync` symbols as they'd show up in docs on angular.io?

**Review by AndrewKushnir — COMMENTED** [review]
Looks great, just added a few comments.

File: goldens/public-api/router/router.md (hunk 1 above)
Comment (atscott) [reply goldens/public-api/router/router.md]: A string redirect isn't a bad idea but does have its own problems. Creating a `UrlTree` from a `string` is relatively easy with `router.parseUrl`. However, it's not always easy to know what the string URL should be, especially when dealing with auxiliary routes. In those cases, it's likely easier to use `Router.createUrlTree` instead, which returns a `UrlTree`.

File: packages/router/src/apply_redirects.ts (hunk 3 above)
Comment (atscott) [reply packages/router/src/apply_redirects.ts]: Given that we have a migration for it, I'm not super concerned. If you feel like it would be beneficial, we could add it.

File: packages/core/schematics/migrations/guard-url-tree/util.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    });
+
+
+    function visitClass(clazz: ts.ClassDeclaration, typeChecker: ts.TypeChecker) {
+      for (const member of clazz.members) {
+        if (!ts.isMethodDeclaration(member) || !methods.some(m => m === member.name.getText())) {
+          continue;
+        }
+        updateMethodReturnType(member);
+        member.forEachChild(function visitNode(node: ts.Node) {
```
Comment (crisbeto) [inline packages/core/schematics/migrations/guard-url-tree/util.ts:34]: I don't think that we should be recursing using `forEachChild` in this case. Consider the following example:

```
canActivate() {
  const foo = () => {
    return 'bar';
  };

  return false;
}
```

I believe that in this case it'll hit the `return 'bar'` before hitting the `return false`.

File: packages/core/schematics/migrations/guard-url-tree/util.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        visitReturnedConditional(statement.expression, typeChecker);
+        return;
+      }
+      visitReturnExpression(statement.expression, typeChecker);
+    }
+
+    function visitReturnedConditional(
+        conditional: ts.ConditionalExpression, typeChecker: ts.TypeChecker) {
+      const whenTrue = conditional.whenTrue;
+      const whenFalse = conditional.whenFalse;
```
Comment (crisbeto) [inline packages/core/schematics/migrations/guard-url-tree/util.ts:84]: nit: you could combine these into `const {whenTrue, whenFalse} = conditional`.

File: packages/core/schematics/migrations/guard-url-tree/util.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      } else if (typeChecker.typeToString(returnType).startsWith('Promise')) {
+        visitPromiseReturnExpression(oldReturn);
+      } else {
+        const assignment = ts.factory.createPropertyAssignment('urlTree', oldReturn);
+        const newReturn = ts.factory.createObjectLiteralExpression([assignment]);
+        results.push({oldNode: oldReturn, newNode: newReturn});
+      }
+    }
+
+    function hasUrlTreeReturnType(type: ts.Type): boolean {
```
Comment (crisbeto) [inline packages/core/schematics/migrations/guard-url-tree/util.ts:113]: Should all of these utility functions be moved outside of `findReturnExpressionsToMigrate`? Otherwise they'll be recreated on each call.

File: packages/core/schematics/migrations/google3/guardUrlTreeRule.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    // Record the changes collected in the import manager.
+    transform.recordChanges();
+
+    if (updateRecorders.has(sourceFile)) {
+      return updateRecorders.get(sourceFile)!.failures;
+    }
+    return [];
+
+    /** Gets the update recorder for the specified source file. */
+    function getUpdateRecorder(sourceFile: ts.SourceFile): TslintUpdateRecorder {
```
[the remark this answers is not in the document]
Comment (atscott) [reply packages/core/schematics/migrations/google3/guardUrlTreeRule.ts:39]: This function, and the one in the `guard-url-tree/index.ts`, read `updateRecorders` which is local to `applyWithProgram` and `runMigration`. I'm going to resolve these two since I don't feel like there's a strong reason to refactor this. But feel free to re-open if you _do_ have strong feelings about it.

File: packages/core/schematics/migrations/guard-url-tree/util.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+const methods = ['canActivate', 'canActivateChild', 'canLoad', 'canDeactivate'];
+
+export function findReturnExpressionsToMigrate(
+    sourceFile: ts.SourceFile, typeChecker: ts.TypeChecker, importManager: ImportManager) {
+  const results: Array<{oldNode: ts.Node, newNode: ts.Node}> = [];
+  (() => {
+    sourceFile.forEachChild(function visitNode(node: ts.Node) {
+      if (ts.isClassDeclaration(node)) {
+        visitClass(node, typeChecker);
```
[the remark this answers is not in the document]
Comment (atscott) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:21]: It's true that this is possible. Do you feel like it is a use-case that's common enough to consider? I generally feel that migrations shouldn't need to attempt to cover the <1% cases.

File: packages/core/schematics/migrations/guard-url-tree/util.ts (hunk 5 above)
Comment (atscott) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:34]: This is accurate. Do you have any suggestions for what you'd like to have instead?

File: packages/core/schematics/migrations/guard-url-tree/util.ts (hunk 7 above)
Comment (atscott) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:113]: That's right, once for each source file. Is this a huge performance concern? The code would otherwise have to pass around the results variable.

File: packages/core/schematics/migrations/guard-url-tree/util.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (crisbeto) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:21]: I don't think so, but I wanted to point it out in case you wanted to handle it. Feel free to resolve this discussion if you don't think it should be handled.

File: packages/core/schematics/migrations/guard-url-tree/util.ts (hunk 5 above)
Comment (crisbeto) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:34]: You could use `method.body.statements` to only walk the top level of statements, although that won't catch a `return` inside of an `if`. Alternatively you could keep your current code, but stop the recursion when you hit an arrow function or a plain `function` declaration.

File: packages/core/schematics/migrations/guard-url-tree/util.ts (hunk 7 above)
Comment (crisbeto) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:112]: I suspect that the engine could optimize this away, but I wanted to flag it since we've been avoiding it everywhere else in the codebase. I don't feel strongly against it.

File: packages/core/schematics/migrations/guard-url-tree/util.ts (hunk 5 above)
Comment (atscott) [reply packages/core/schematics/migrations/guard-url-tree/util.ts:34]: Yea, I was definitely trying catch returns inside control flows. This is a super common case for guards. I'll adjust it to stop the recursion at functions.

File: goldens/public-api/router/router.md (hunk 1 above)
Comment (AndrewKushnir) [reply goldens/public-api/router/router.md]: I was just thinking about simple use-cases, like checking if a user is logged in and redirect if not ([what we have in examples here](https://github.com/angular/angular/pull/45023/files#diff-749c533e8f68f708dab9bd7b718e04090cde7040be5cc65dcae4fb90e98baf02R47)). Here is a simplified version of that example:

```
  checkLogin(url: string): true|RedirectCommand {
    if (this.authService.isLoggedIn) { return true; }

    // Redirect to the login page with extras
    return {redirectTo: this.router.createUrlTree(['/login'])};
  }
```
If we have a string version of the URL to go to, it may look like this:
```
  checkLogin(url: string): true|RedirectCommand {
    if (this.authService.isLoggedIn) { return true; }

    // Redirect to the login page with extras
    return {redirectTo: '/login'};
  }
```
This also reminds me of the `redirectTo` field from the `Route` interface, which is also a string: https://angular.io/api/router/Route#redirectTo (it's not the same thing, but semantics is very similar).

I understand that there may be cons of doing that and current guards already return `UrlTree`, so this particular API change would not complicate the surface (so I'm ok if we keep it as implemented). I just thought it'd be cool if for common use-cases there is a simple way to express the redirect (that would also help us to explain guards to users in various docs on the website), while also having a full power of configuration (through constructing a URL with extra options and control how the redirect should happen) when it's needed (which looks less common in general?).

File: packages/router/src/apply_redirects.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/router/src/apply_redirects.ts]: Yeah, I was thinking about possible backsliding that may happen in apps code where people would use an old pattern that they've used to. In this case there is no indication except the `@deprecated` tag (may be that's sufficient though if it shows up in the IDE as a hint?).

**Review by AndrewKushnir — DISMISSED** [review]
Looks great, thanks @atscott 👍 

~~I'd be curious to see what other `public-api` group reviewers think about https://github.com/angular/angular/pull/45023#discussion_r803090836, but I'm ok to proceed with the current implementation and add a simple (string-based) `redirectTo` support later if needed.~~ Update: we can extend the logic later if needed, so this question is not relevant at this moment.

File: goldens/public-api/router/router.md (hunk 1 above)
Comment (atscott) [reply goldens/public-api/router/router.md]: I do understand the use-case but I think it's actually a lot more complicated than you want to believe :) `Route.redirectTo` has plenty of limitations and confusing behaviors. I wouldn't not want those same semantics to be carried over to a new API. Here is a non-exhaustive list of some reasons I don't think `string` should be available here:

1. Is the string required to be an absolute navigation only? (i.e. start with a `/`)
2. If the above is true, this is now different from the requirements on `Route.redirectTo`. That's kind of confusing.
3. If #1 is _not_ true, the redirect would have to be computed relative to the `ActivatedRouteSnapshot` at the location of the guard. This is a bit confusing for an API because the redirect would then depend on the location of the guard and sort of reach "outside" of the guard's implementation boundaries. That is, it's highly dependent on how the overall router is designed right now
4. `{redirectTo: this.router.parseUrl('/login')}` is not significantly more complicated than `{redirectTo: '/login'}`
5. The same questions will exist around how or if query params (and route params) should be preserved that exist today for `Route.redirectTo`


In short, if we were to expand the type to allow a `string`, we would have to consider more than just the simple case that you have above. This isn't necessarily to say we can't or shouldn't do this later. Now that we have a nice `RedirectCommand` interface, expanding the `redirectTo` property to include `string` at a later date _can_ be done but should _only_ be done after very careful consideration of all use-cases. `UrlTree` is very straightforward in comparison.

Comment (AndrewKushnir) [reply goldens/public-api/router/router.md]: Great summary, thanks Andrew! 👍

I'm marking this as resolved and happy to proceed with the current implementation and consider extending it later if needed.

**Review by crisbeto — DISMISSED** [review]
LGTM

**Comment by atscott** [comment]
Hello kind public API reviewers. There have been a few updates to this PR that warrants a re-review:

1. The `RedirectCommand` is now a class instead of an interface, making it easier to determine if the value is a redirect with `instanceof`
2. The change above allows for resolvers to return a `RedirectCommand`, which is an extra commit on this PR
3. I've dropped the migration from the PR to make it easier to review and land. I'll do the migration as a follow-up instead.

**Comment by atscott** [comment]
@rhutchison I do still see value in this but work on it has dropped pretty significantly off my priority list. I would say the deprecation and migration of returning just `UrlTree` was/is the biggest blocker due to the time consumption and churn it would create. I might pick this up again as a feature-only change that removes the deprecation, at least for the time-being.

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Comment by atscott** [comment]
TGP=https://fusion2.corp.google.com/presubmit/583045890/OCL:583045890:BASE:614718613:1710212556271:4b5e09e3/targets

**Comment by atscott** [comment]
This PR was merged into the repository by commit 8735af08b976b30cf236a83f9e8b64b5ff62e9f3.
